Link Project to Repository or Team Command

benebsiny opened the pull request:
<!--
  Thank you for contributing to GitHub CLI!
  To reference an open issue, please write this in your description: `Fixes #NUMBER`
-->

Fixes #7849 
This PR implements `gh project link` command. Here are some examples:
```bash
# Link project to a repo
gh project link 1 --repo my-repo

# Link project to a team
gh project link 1 --team my-team

# Unlink project from a repo
gh project unlink 1 --repo my-repo

# Unlink project from a team
gh project unlink 1 --team my-team
```

benebsiny committed:
Link Project to Repository or Team Command

benebsiny committed:
Remove support for team

benebsiny committed:
Support for unlink

benebsiny committed:
Support for team

benebsiny committed:
Add testing for link and unlink

benebsiny committed:
Fix wording

benebsiny committed:
Merge branch 'trunk' into cli-7849

andyfeller commented on pkg/cmd/project/link/link.go:
```diff
@@ -0,0 +1,195 @@
+				io:         f.IOStreams,
+			}
+
+			if config.opts.repo != "" && config.opts.team != "" {
+				return fmt.Errorf("specify only one repo or team")
+			} else if config.opts.repo == "" && config.opts.team == "" {
+				return fmt.Errorf("specify at least one repo or team")
+			}
```
Formatting the flags to match other commands and rewording the latter case to avoid any confusion that multiple `--repo` or `--team` flags can be specified.

```suggestion
			if config.opts.repo != "" && config.opts.team != "" {
				return fmt.Errorf("specify only one of `--repo` or `--team`")
			} else if config.opts.repo == "" && config.opts.team == "" {
				return fmt.Errorf("specify either `--repo` or `--team`")
			}
```

williammartin replied:
Sorry I had a review in draft about this already. Preferably use `cmdutil.MutuallyExclusive` in this case to ensure we are standardised and pass up a flag error.

andyfeller commented on pkg/cmd/project/link/link.go:
```diff
@@ -0,0 +1,195 @@
+	}
+
+	linkCmd.Flags().StringVar(&opts.owner, "owner", "", "Login of the owner. Use \"@me\" for the current user.")
+	linkCmd.Flags().StringVarP(&opts.repo, "repo", "R", "", "The repository to be linked to this project")
+	linkCmd.Flags().StringVarP(&opts.team, "team", "T", "", "The team to be linked to this project")
```
Generally want to default to lowercase flags to avoid of extra keystrokes but also `-R,--repo  [HOST/]OWNER/REPO` is a standard flag on nearly every other command.

```suggestion
	linkCmd.Flags().StringVarP(&opts.repo, "repo", "r", "", "The repository to be linked to this project")
	linkCmd.Flags().StringVarP(&opts.team, "team", "t", "", "The team to be linked to this project")
```

benebsiny replied:
@andyfeller There is already a shorthand `t` for `--template`😱

williammartin replied:
Ack, I went through this discovery as well, `-R` and `-T` are fine for me.

Dynasty-coder replied:
linkCmd.Flags().StringVarP(&opts.team, "team", "T", "", "The team to be linked to this project
```

andyfeller commented on pkg/cmd/project/link/link.go:
```diff
@@ -0,0 +1,195 @@
+		}
+		return printResults(config, query.LinkProjectV2ToTeam.Team.URL)
+	}
+	return fmt.Errorf("specify at least one repo or team")
```
If we are handling this in `RunE`, then we can just return nil here:

```suggestion
	return nil
```

andyfeller commented on pkg/cmd/project/link/link.go:
```diff
@@ -0,0 +1,195 @@
+		}
+		config.opts.repoID = repo.ID
+
+		query, variable := linkRepoArgs(config)
+		err = config.client.Mutate("LinkProjectV2ToRepository", query, variable)
+		if err != nil {
+			return err
+		}
+
+		if config.opts.exporter != nil {
+			return config.opts.exporter.Write(config.io, query.LinkProjectV2ToRepository.Repository)
+		}
+		return printResults(config, query.LinkProjectV2ToRepository.Repository.URL)
+
+	} else if config.opts.team != "" {
+		teams, err := api.OrganizationTeams(c, ghrepo.New(owner.Login, ""))
+		if err != nil {
+			return err
+		}
+		for _, team := range teams {
+			if team.Slug == config.opts.team {
+				config.opts.teamID = team.ID
+				break
+			}
+		}
+		if config.opts.teamID == "" {
+			return fmt.Errorf("can't find team %s", config.opts.team)
+		}
+
+		query, variable := linkTeamArgs(config)
+		err = config.client.Mutate("LinkProjectV2ToTeam", query, variable)
+		if err != nil {
+			return err
+		}
+
+		if config.opts.exporter != nil {
+			return config.opts.exporter.Write(config.io, query.LinkProjectV2ToTeam.Team)
+		}
+		return printResults(config, query.LinkProjectV2ToTeam.Team.URL)
+	}
```
Moving the bulk of these 2 conditional blocks to separate functions would help with comprehension and testing.  Not a deal breaker but this is on the edge of a lot for a single function with multiple code paths.

Dynasty-coder replied:
R,--repo  [HOST/]OWNER/REPO

andyfeller commented on pkg/cmd/project/unlink/unlink.go:
```diff
@@ -0,0 +1,195 @@
+				io:         f.IOStreams,
+			}
+
+			if config.opts.repo != "" && config.opts.team != "" {
+				return fmt.Errorf("specify only one repo or team")
+			} else if config.opts.repo == "" && config.opts.team == "" {
+				return fmt.Errorf("specify at least one repo or team")
+			}
```
Bringing this inline with the suggestion above:

```suggestion
			if config.opts.repo != "" && config.opts.team != "" {
				return fmt.Errorf("specify only one of `--repo` or `--team`")
			} else if config.opts.repo == "" && config.opts.team == "" {
				return fmt.Errorf("specify either `--repo` or `--team`")
			}
```

andyfeller commented on pkg/cmd/project/unlink/unlink.go:
```diff
@@ -0,0 +1,195 @@
+	}
+
+	linkCmd.Flags().StringVar(&opts.owner, "owner", "", "Login of the owner. Use \"@me\" for the current user.")
+	linkCmd.Flags().StringVarP(&opts.repo, "repo", "R", "", "The repository to be unlinked from this project")
+	linkCmd.Flags().StringVarP(&opts.team, "team", "T", "", "The team to be unlinked from this project")
```
```suggestion
	linkCmd.Flags().StringVarP(&opts.repo, "repo", "r", "", "The repository to be unlinked from this project")
	linkCmd.Flags().StringVarP(&opts.team, "team", "t", "", "The team to be unlinked from this project")
```

andyfeller commented on pkg/cmd/project/link/link.go:
```diff
@@ -0,0 +1,195 @@
+		return printResults(config, query.LinkProjectV2ToRepository.Repository.URL)
+
+	} else if config.opts.team != "" {
+		teams, err := api.OrganizationTeams(c, ghrepo.New(owner.Login, ""))
+		if err != nil {
+			return err
+		}
+		for _, team := range teams {
+			if team.Slug == config.opts.team {
+				config.opts.teamID = team.ID
+				break
+			}
+		}
+		if config.opts.teamID == "" {
+			return fmt.Errorf("can't find team %s", config.opts.team)
+		}
```
It seems like there should be a better way to find 1 team based on slug in an organization 🤔

benebsiny replied:
I've created a new API to retrieve the ID of a team directly.

andyfeller reviewed (dismissed):
@benebsiny: wow!  thank you for the contribution including tests! 🤗 

While @williammartin captures his thoughts, I thought I'd do an initial once over for a few things that stood out to me.  All in all, appreciate your efforts!

benebsiny committed:
fix wording, move conditional blocks to functions, and add api to retrieve the ID of a team

williammartin commented on pkg/cmd/project/link/link.go:
```diff
@@ -0,0 +1,194 @@
+	config.opts.repoID = repo.ID
+
+	query, variable := linkRepoArgs(config)
+	err = config.client.Mutate("LinkProjectV2ToRepository", query, variable)
```
How would you feel about creating a `LinkProjectToRepository` on to the `client`? Looking at `Mutate` it looks like it leaks the GQL implementation to the caller (the doc indicates some tech debt here and I agree).

The same for `LinkProjectToTeam` and the `Unlink` variants.

benebsiny replied:
@williammartin I moved the functions to the `client`, take a look at the latest changes 🚀

benebsiny committed:
handle flags with `cmdutil.MutuallyExclusive`, and move mutation methods to `queries.go`

williammartin commented on pkg/cmd/project/link/link_test.go:
```diff
@@ -0,0 +1,345 @@
+	client := queries.NewTestClient()
+
+	// get repo ID
+	httpReg := &httpmock.Registry{}
+	defer httpReg.Verify(t)
+
+	httpReg.Register(
+		httpmock.GraphQL(`query RepositoryInfo\b`),
+		httpmock.StringResponse(`{"data":{"repository":{"id": "repo-ID"}}}`),
+	)
+	httpClient := newTestClient(httpReg)
```
Can we get rid of this and do something like:

```
	gock.New("https://api.github.com").
		Post("/graphql").
		MatchType("json").
		JSON(map[string]interface{}{
			"query": "query OrganizationTeam.*",
		}).
		Reply(200).
		JSON(map[string]interface{}{
			"data": map[string]interface{}{
				"organization": map[string]interface{}{
					"team": map[string]interface{}{
						"id": "team-ID",
					},
				},
			},
		})
```

And then pass the `http.DefaultClient` in the factory function below. I tried this out and it seems to work, but I want to give you a chance to look at it.

It seems like we should try to settle on just one form of http mocking.

williammartin replied:
Same for the other tests in `link` and `unlink` of course.

williammartin commented on pkg/cmd/project/link/link.go:
```diff
@@ -0,0 +1,162 @@
+}
+
+func linkTeam(c *api.Client, owner *queries.Owner, config linkConfig) error {
+	team, err := api.OrganizationTeam(c, ghrepo.New(owner.Login, ""), config.opts.team)
```
`ghrepo.New(owner.Login, "")` doesn't look quite right to me. This is going to construct a new `Repo` with the default host (`github.com`). This host is then used in the `OrganizationTeam` function as the URL to send the GQL query to. However, it's possible that we're supposed to be targeting an Enterprise Server installation e.g. `ghe.io`.

I think instead what we should do here is make the signature: `OrganizationTeam(client *Client, hostname string, org string, teamSlug string)` and pass exactly what is needed (note that I think we should do work to make this `client.OrganizationTeam(org, teamSlug string)` but that's out of scope for now). I know you copied from the surrounding function signatures but tbh they are not good. There's absolutely no reason for a function that fetches teams for projects for an organization to take a `repo` as an argument. This time it's leaking from the caller into the function because the callers happened to have a `repo` to hand.

With that signature change, we need to find the `host` that we should be using and I would suggest we use `host, _ := cfg.Authentication().DefaultHost()` and pass that `host` in, as exists in other functions like `gh gist list`.

Let me know if this makes sense or not, it's kind of meta.

benebsiny replied:
I agree with you. My initial idea was to keep the parameters of functions in the same file as consistent as possible, even if it looks a bit odd. However, with the question you've raised, I think implementing it your way would be better

benebsiny replied:
Should the `host` in `linkRepo` function be retrieved in the same way?

williammartin replied:
@benebsiny I think you added a question and then deleted it cause I can't see it anymore but actually I think it was a really good question. Indeed I think for `linkRepo` we also need to allow the opportunity for the repo to be on different hosts.

I also have a question, if I were in a repo say `cli/cli` and I ran `gh project link 1` would you expect it to work? What about `GH_REPO=cli/cli gh project link 1`?

The reason I'm asking this is because other commands that work with a repo support this and the approach to solving `linkRepo` above will vary based on this design decision.

benebsiny replied:
> I think you added a question and then deleted it cause I can't see it anymore...

Yes, I asked the question, and I later found the answer (the answer is same as yours), so I deleted that question.

> I also have a question, if I were in a repo say cli/cli and I ran gh project link 1 would you expect it to work?

I think it's really a good feature. I can implement it.

williammartin replied:
Yeh so in that regard I think you should look at something like the `issue` command and the `EnableRepoOverride` function. Sorry I don't have time to provide more guidance right now but see how it is used for `issue view` (or any other commands)

williammartin commented on pkg/cmd/project/link/link.go:
```diff
@@ -0,0 +1,162 @@
+	return printResults(config, result.URL)
+}
+
+func printResults(config linkConfig, url string) error {
```
Is there something that you wanted to get out of the `Team` or `Repository` URL, or wanted to have scripted with the exporter? I'm not sure what the workflow is that would make this useful information to have and I'm always in favour of removing code unless we have a specific need.

benebsiny replied:
Just like other `gh projects` commands, print out the URL upon successful execution. Should we refrain from printing anything after a successful execution?

williammartin replied:
I'm not sure. I think if we were to print something it should be like `"Linked X to Y" and "Unlinked X from Y". What do you think?

benebsiny replied:
It's good. It looks intuitive.

benebsiny committed:
Replace `httpmock` with `gock` in testing

benebsiny committed:
Refactor `OrganizationTeam`

benebsiny committed:
Change the output of link` and `unlink`

benebsiny committed:
Support for directory based repo and `GH_REPO` env variable

benebsiny committed:
Fix printed project number

Fix the error regarding the printed error number when selecting the project number in interactive mode

benebsiny commented:
@williammartin The support for directory-based repo and `GH_REPO` has been implemented, checkout the latest changes 🚀

williammartin commented:
This is looking very close 🎉 . I've played around with the UX today and I think there's just a few things that could do with some love.

Firstly, adding some text to the help description and example to both commands demonstrating that if no `--team` or `--repo` is provided, it will attempt to use the base repo. I think as a follow up we should consider making the error message better if neither `--repo` or `--team` are provided **and** we're not in a git repo, but I think that's a problem across the board with commands that expect to be in a repo so I don't expect you to handle it here.

Secondly, I really don't think that the `json` exporter is providing much value here. Here's what it looks like for `link`ing a team and repo:

```
➜  gh project link --format json --owner williammartin-test-org --team a-team 1
{
  "ID": "T_kwDOB-vozM4AeSfB",
  "URL": "https://github.com/orgs/williammartin-test-org/teams/a-team"
}
```

```
➜  gh project link --format json --owner williammartin-test-org --repo test-repo 1
{
  "Name": "test-repo",
  "ID": "R_kgDOJgN1OA",
  "URL": "https://github.com/williammartin-test-org/test-repo"
}
```

I don't find it to be particularly useful information. It's also not necessarily the information they might want to see, maybe they want to see the project data rather than the repository or team? I don't think we should be tying our CLI UX to an individual query like this. If we remove this then we can also trim down our graphql data transfer because we no longer need to ask for the `Team` or `Repository` as part of the mutation.

If people ask for this data to be machine readable in the future then we can easily add it back without handcuffing ourselves into supporting it forever now. Make sense?

benebsiny commented:
@williammartin It makes sense not to print the `Team` or `Repository` as a result. Maybe the `--format json` option can be removed for now. Thanks for your feedback!

williammartin commented:
Yeh I think I would just remove `--format`. I like the human readable text that is displayed though. It might even be useful if we have the project name to display that too?

benebsiny commented:
@williammartin How about `project #1 'project-name'`?

williammartin commented:
Sounds good

benebsiny committed:
Remove `--format` flag and add description for not specifying `--repo` and `--team` flag

benebsiny commented:
@williammartin I've made some changes, checkout the latest updates. Thanks!

williammartin reviewed (changes requested):
Sorry it's been a little while, I've been out sick.

I got a little confused by one thing, which is that:

```
➜ ./bin/gh project link --repo williammartin-test-org/test-repo --owner williammartin-test-org 1
GraphQL: Could not resolve to a Repository with the name 'williammartin-test-org/williammartin-test-org/test-repo'. (repository)
```

Doesn't work. I think you've made a design decision here that you shouldn't need to provide the org part because it can be inferred by the owner but I think we should be flexible in this case because it's how `-R` is used everywhere else. If they provide `org-A/repo` and `--owner org-B` then they should get a useful error that this won't work (at least, right now).

If you're getting tired of making these little changes I'd be happy to do them for you but I also want to give you the opportunity to learn and own it yourself.

I'd also request as a minor thing that anywhere you use `assert.NoError` or `assert.Error` that you replace them with the `require.NoError` and `require.Error` equivalents. It's almost never useful to see the output of other assertions after an error and is often confusing.

benebsiny commented:
@williammartin I have an idea. How about eliminating the use of `--owner` and instead using `--repo owner/repo` or `--team org/team`? This approach is more common and is less likely to confuse users

---
(The comment that was just deleted)
Should the `assert.Equal` also be replaced with `require.Equal`?

benebsiny commented:
Should the `assert.Equal` also be replaced with `require.Equal`?

benebsiny committed:
Merge branch 'trunk' of github.com:cli/cli into trunk

benebsiny committed:
Replace `assert` with `require` in testing

williammartin commented:
> I have an idea. How about eliminating the use of --owner and instead using --repo owner/repo or --team org/team? This 
> approach is more common and is less likely to confuse users

What I'm not sure about here is that all the other `gh project` commands take the `--owner` flag to help disambiguate. It seems unfortunate to change that.

Do you think there is an issue with supporting both `--repo owner/repo` and `--repo repo` (and the same for team)?

> Should the assert.Equal also be replaced with require.Equal?

I prefer this, but I don't mind. `require.Error` and `require.NoError` are important because it almost never makes sense to see the result of assertions after them because it means something has gone extremely wrong in the test and the following assertions will just be noise. Sometimes it makes sense to see the values of other assertion failures together.

benebsiny commented:
> Do you think there is an issue with supporting both --repo owner/repo and --repo repo (and the same for team)?

No, I can try to implement this feature

benebsiny committed:
Support for `[HOST/]OWNER/REPO` format

benebsiny committed:
chore: simplify parameter `owner`

benebsiny commented:
@williammartin I have made some changes, take a look at the latest commit!

williammartin reviewed (approved):
LGTM!

There is one minor thing which I would welcome a follow up for but I'm not going to block this PR on and that is the following case:

```
Given I in a cloned git repository my-org/my-repo
When I run gh project link --owner other-org
Then I would expect an error message that the orgs do not match
```

Thank you for all your hard work and back and forth here, and for improving the CLI!

williammartin commented:
I'm going to make an assumption here that Andy's original changes requested were addressed because it was a long time ago and they were addressed either with a different approach or pointing out reason for uppercase.
